features/mbedtls/README.md

```diff
@@ -1,14 +1,96 @@
+
+3. [**Hashing:**](https://github.com/ARMmbed/mbed-os-example-tls/tree/master/hashing) Hashing demonstrates the various APIs for computing hashes of data (also known as message digests) with SHA-256.
+
+4. [**Authenticated encryption:**](https://github.com/ARMmbed/mbed-os-example-tls/tree/master/authcrypt) Authcrypt demonstrates usage of the Cipher API for encrypting and authenticating data with AES-CCM.
+
+
+These examples are fully integrated into Mbed OS. Each of them comes with complete usage instructions as a `README.md` file in the directory of each example.
+
+
+Configuring Mbed TLS features
+-----------------------------
+
+Mbed TLS makes it easy to disable any feature during compilation, if that feature isn't required for a particular project. The default configuration enables all modern and widely-used features of the TLS protocol, which should meet the needs of most projects. It also disables all older and less common features, to minimize the code footprint.
+
+The list of available compilation flags is available in the fully documented [config.h file](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/config.h).
+
+If you need to adjust those flags, you can provide your own supplementary configuration-adjustment file with suitable `#define` and `#undef` statements. These will be included between the default definitions and the sanity checks. Your configuration file should be in your application's include directory, and can be named freely; you just need to let mbed TLS know the file's name. To do that, you can use the [Mbed OS Configuration
+system](https://docs.mbed.com/docs/mbed-os-api/en/latest/api/md_docs_config_system.html)
+
+For example, if you wanted to enable the options, `MBEDTLS_PEM_WRITE_C` and `MBEDTLS_CMAC_C`, and provide your own additional configuration file for Mbed TLS named `my_config.h`, you could define these in a top level `mbed_app.json` configuration file in the root directory of your project.
+
+The Mbed TLS configuration file would be specified in the `.json` file as follows.
+
+```
+{
+    "macros" : [
+
+        "MBEDTLS_USER_CONFIG_FILE" : "my_config.h",
+
+        "MBEDTLS_PEM_WRITE_C",
+        "MBEDTLS_CMAC_C"
+    ]
+    ....remainder of file...
+}
+```
+
+The additional configuration file, `my_config.h`, can then be used as a normal configuration header file to include additional configurations. For example, it could include the follow lines to include ECJPAKE, and the CBC block mode:
+
+```
+    #define MBEDTLS_ECJPAKE_C
+    #define MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED
+
+    #undef MBEDTLS_CIPHER_MODE_CBC
+```
+
+## Getting Mbed TLS from GitHub
+
+Mbed TLS is maintained and developed in the open, independently of Mbed OS, and its source can be found on GitHub here: [ARMmbed/mbedtls](https://github.com/ARMmbed/mbedtls). As a consequence, unlike other parts of mbed OS, changes to Mbed TLS must be committed upstream in the Mbed TLS repository.
```

Comment:
I find this section a bit confusing. “Changes to Mbed TLS must be committed upstream” — so this is about contributing to Mbed OS? But the rest of the section only discusses retrieving some different version of Mbed TLS, not contributing to it. Why am I reading this anyway? What problem does this section help me solve?

I think if you drop the “As a consequence …” sentence and join the previous sentence with the next paragraph, the content of the section is fine. And the title should be more explicit: “Switching to a different version of Mbed TLS” or s/different/new/.

Reply:
As written, I agree that the "As a consequence" sentence is a bit badly targeted and a bit of an aside. What's being said is worth saying somewhere, although I'm not sure the best place for it.

Essentially, we want to make tell people that if they want to modify any Mbed TLS bits that are present in the Mbed OS repository, those changes have to be done upstream. That essence doesn't fit within a "Getting Mbed TLS from GitHub" or a "Switching to a different version of Mbed TLS" section.